unattended-upgrade

```diff
@@ -204,9 +228,44 @@ class UnattendedUpgradesCache(apt.Cache):
         if function == apt.package.Package.mark_upgrade \
            and not pkg.is_upgradable:
             raise NoAllowedOriginError
-        function(pkg, **kwargs)
-        changes = self.get_changes()
-        for marked_pkg in changes:
+        marking_succeeded = self.call_checked(function, pkg, **kwargs)
+        if not marking_succeeded:
+            # The package could not have been marked to install/upgrade
+            # with the current candidate adjustments.
+            # Try marking it without the adjustments then adjust
+            # all packages to be installed/upgraded. This will work where
+            # package dependencies are similar for the adjusted and not
+            # adjusted case, just with different versions.
+            logging.debug("falling back to marking %s, then adjusting changes"
+                          % pkg.name)
+            for pkg2 in ([self[pkg_name] for pkg_name
+                          in self._cached_candidate_pkgnames]):
+                pkg2.candidate = pkg2.versions[0]
```

Comment:
Um, why is this resetting all candidates to to their 0th version (which is IIRC, always the latest available version)? That seems wrong.

Reply:
@julian-klode Typically the highest versions are installable with the highest versions of the dependencies, thus the package set would be pulled from -updates. Later if that work we try to adjust the set to versions from allowed origins and if that worked we are ok.
The reason of that is sudo apt install systemd=237-3ubuntu10.13 fails because .15 is available from updates and needs strict versioned dependencies.

Reply:
OK